Modules/_sre.c

```diff
@@ -427,11 +427,15 @@ state_init(SRE_STATE* state, PatternObject* pattern, PyObject* string,
     }
 
     /* adjust boundaries */
+    if (start < 0)
+        start += length;
     if (start < 0)
         start = 0;
```

Comment:
In `re.compile('.').findall("abcd", -10, -3)`, the first `if` will set `start` to `-6` which is not a valid index, and the second if is necessary because it will set it to `0`.  There are currently no tests for this scenario, so I would add a couple.